specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+CARv2 is a minimal upgrade to the [CARv1](../carv1/index.md) format with the primary aim of adding an optional index within the format for fast random-access to blocks.
+
+CARv2 makes use of CARv1 by wrapping a properly formed CARv1 with a prefix containing a pragma and header, and a suffix containing the optional index data. Once the offset and length of the CARv1 bytes are determined using CARv2 parsing rules, an existing CARv1 decoder can be used to read the roots and `CID:Bytes` pairs. Likewise, a CARv1 encoder can be used to encode this data for wrapping by a CARv2 encoder.
+
+## Format Description
+
+CARv2 consists of:
+
+1. An 11-byte pragma (or sequence of "magic bytes") that identify the data as a CARv2 format.
+2. A header describing data payload offset and length and index offset within the container as well as a bitfield for setting various capabilities supported by the container.
+3. A standard CARv1 data payload, including standard CARv1 header and roots and sequence of `CID:Bytes` pairs.
+4. An optional index payload, which may be one of a number of supported index formats, allowing for fast lookups of blocks within the data payload in the form of an offset from the start of the container.
+
+The CARv2 format can be illustrated as follows:
+
+```
+| 11-byte fixed pragma | 40-byte header | optional padding | CARv1 data payload | optional padding | optional index payload
+```
+
+### Pragma
+
+The CARv2 version pragma (or "magic bytes") was chosen for compatibility with existing CARv1 parsers. CARv1 leads with a [DAG-CBOR](../../../codecs/dag-cbor/index.md) block, prefixed with a varint where the block contains the version and roots array:
+
+```ipldsch
+type CarV1Header struct {
+  version Int # 1
+  roots [&Any]
+}
+```
+
+To introduce a new version that existing parsers can safely reject as an *"unsupported version"* we must copy the minimal form of this header with a new version.
+
+Therefore, the CARv2 header is a fixed sequence of 11 bytes: **`0x0aa16776657273696f6e02`**.
+
+These bytes decode as follows: A leading `0x0a` which translates as a `uint(10)` (or `varint(10)`) indicating the length of the DAG-CBOR header block to follow. The remaining 10 bytes are a standard CBOR encoding of a map containing a `"version"` field with a value `2`. i.e.
+
+
+```ipldsch
+type CarV2Header struct {
+  version Int # 2
+}
+```
+
+The 10 bytes are parsed in CBOR as follows:
+
+```
+a1                                                # map(1)
+  67                                              #   string(7)
+    76657273696f6e                                #     "version"
+  02                                              #   uint(2)
+```
+
+Existing CARv1 parsers should safely read this pragma and reject the container as an unsupported version of the CAR format.
+
+This 11 byte string remains fixed and may be matched using a simple byte comparison and does not require a varint or CBOR decode since it does not vary for CARv2 containers.
```

Comment:
> containers

Would the word "archives" be a better fit?

Reply:
Not really, cause it'd be "Content Addressable aRchives archives". I could just try stripping out the descriptor though. I'm consciously avoiding "file" here, same as the v1 spec, which is also why this spec is in a section called "transports". In the v1 I didn't use "container" and stuck to either just saying "CAR" or "CAR format".

I've removed all of the uses of "container" here, I'm not sure if it's an improvement, however, it makes heavy use of "CARv2" as a noun now. Let me know what you think.

Reply:
That makes sense thank you